### ISSUE TITLE
Session.execute propagates **kw to Session.get_bind which won't take them

### ISSUE THREAD

**Issue by tobiasherp** [sqlalchemy/sqlalchemy#5573]
**Describe the bug**

I use `sqlalchemy.orm.Session.execute` to execute simple SQL statements I already have.
When I specify query data (the `**kw` of the `execute` method), I get a `TypeError`, because the `**kw` are passed on to the `get_bind` method of the same class (unless I specify the `bind` argument as well; but this should be optional).

**Expected behavior**

When the `bind` argument is *None*, the `execute` method should compute it itself by correctly calling the `get_bind` method.

**To Reproduce**
Please try to provide a [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) example. 
See also [Reporting Bugs](https://www.sqlalchemy.org/participate.html#bugs) on the website, and some [example issues](https://github.com/sqlalchemy/sqlalchemy/issues?q=label%3A%22great+mcve%22)

Pardon me for my example not being complete; but the details really don't matter:
The call (in line 1289, as of v1.3.18) to the `get_bind` method (defined in line 1450) must fail, from looking at the signature,
if query data is specified (`**kw`) and `bind` is not.

```py
clause = 'SELECT p2_result, result_label FROM my_table WHERE project_id = :project_id;'
query_data = {'project_id': 42}
connection.execute(clause, **query_data)
```

**Error**

```
Traceback (innermost last):
  ... (outer code from my application)
  Module visaplan.zope.reldb.legacy, line 94, in select
  Module sqlalchemy.orm.session, line 1289, in execute
TypeError: get_bind() got an unexpected keyword argument 'project_id'
```
The solution should be easy: the `execute` method shouldn't use the `kw` query data when calling `get_bind`.

**Versions.**
 - OS: CentOS Linux
 - Python: 2.7
 - **SQLAlchemy: 1.3.18**
 - Database: PostgreSQL
 - DBAPI: psycopg2

**Additional context**

**Have a nice day!**

**Comment by tobiasherp** [comment]
I just noticed that we have version **1.3.19** meanwhile.
I tried that version, and the problem is still there; all information is still correct, including the line numbers.

**Comment by zzzeek** [comment]
Hi there - 

I had a long response here to try to clarify what's being asked, then I realized that i think you are confusing the term "bind arguments used by the Session as the parameters in your SQL query, such as "project_id".   So, I can see the names here are confusing but in SQLAlchemy the term "bind" usually refers to which Engine is going to be used for a certain operation, so in a Session, the "bind arguments" means special arguments that will be used to figure out which Engine should be used.   The parameters for the SQL string are passed as a single argument "params" that is a dictionary of parameters (note below I am also making the leap that when you typed "connection", you meant "session"; an MCVE would have made this much easier to understand what you were asking:

    clause = 'SELECT p2_result, result_label FROM my_table WHERE project_id = :project_id;'
    query_data = {'project_id': 42}
    result = session.execute(clause, query_data)

So it is also very important to note that this whole API that is admittedly not very good is going to be completely refined for SQLAlchemy 2.0, and the next major release 1.4 will contain all of the following changes with deprecation paths for the previous styles, this is all already committed:

* All .execute() methods, that is connection.execute(), session.execute(), will accept the SQL parameters as a single "params" argument which is either a dictionary or sequence of dictionaries.  being able to say conn.execute(stmt, **kwargs) is removed in 2.0

* the Session.execute() **kw arguments which have confused you here are deprecated; the special "bind arguments" are now passed as a single parameter "bind_arguments" to Session.execute() and usually people won't deal with them as they are seldom used

hope this helps!

**Comment by tobiasherp** [comment]
Yes, this helps; thank you.
I'm not sure where I got the idea from that **kw would be the params, and *args would be a list of dicts.

However, given the `**kw` are *deprecated*, they shouldn't make the whole thing crash, right?
So, the suggested change of line 1289 won't hurt, IMO.

**Comment by zzzeek** [comment]
You have to be more specific as to "the change of line 1289" can you illustrate a patch ?

**Comment by zzzeek** [comment]
oh.  dont use the kw.  no can do, the contract of that method in 1.3 is that it is extensible.  take a look at https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_3/lib/sqlalchemy/ext/horizontal_shard.py#L250 which makes use of this.

all fixed in 1.4, no worries

**Comment by zzzeek** [comment]
see https://docs.sqlalchemy.org/en/13/orm/session_api.html?highlight=session%20execute#sqlalchemy.orm.session.Session.execute.params.**kw

**Comment by tobiasherp** [comment]
> oh. dont use the kw.

Understood.

> the contract of that method in 1.3 is that it is extensible. take a look at https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_3/lib/sqlalchemy/ext/horizontal_shard.py#L250 which makes use of this.

Yes, but `.orm.session.Session.get_bind` is / was *not* extensible; it simply takes / took `mapper` and `clause` options, and nothing else.

> all fixed in 1.4 (well, deprecated), no worries

So you have changed that call to `.get_bind` in `sqlalchemy/orm/session.py:1289` and removed the `**kw`, right?
(because *that* was my suggested change)

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Emit deprecation warning for **kw passed to session.execute()** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2222

**Comment by zzzeek** [comment]
> > oh. dont use the kw.
> 
> Understood.
> 
> > the contract of that method in 1.3 is that it is extensible. take a look at https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_3/lib/sqlalchemy/ext/horizontal_shard.py#L250 which makes use of this.
> 
> Yes, but `.orm.session.Session.get_bind` is / was _not_ extensible; it simply takes / took `mapper` and `clause` options, and nothing else.

this is correct, because the default Session does not support additional arguments for binds.   if you subclass Session, you can add new arguments to your get_bind() method.    then you can pass these arguments through when you call session.execute().   this is standard Python **kw stuff.

> 
> > all fixed in 1.4 (well, deprecated), no worries
> 
> So you have changed that call to `.get_bind` in `sqlalchemy/orm/session.py:1289` and removed the `**kw`, right?

it's not removed right now, it's marked as deprecated.  I do need to add a deprecation warning at runtime so thanks for reminding me.

> (because _that_ was my suggested change)
